decision_service/dsclientlib/logger/ds_async_batcher.h

```diff
@@ -0,0 +1,120 @@
+#pragma once
+
+#include "ds_concurrent_queue.h"
+#include "ds_api_status.h"
+
+#include <memory>
+#include <string>
+#include "../error_callback_fn.h"
+#include <thread>
+
+namespace decision_service {
+  class error_callback_fn;
+
+	// This class takes uses a queue and a background thread to accumulate events, and send them by batch asynchronously.
+	// A batch is shipped with TSender::send(data)
+	template <typename TSender>
+	class async_batcher {
+
+	public:
+
+		int append(const std::string& evt, api_status* status = nullptr)
+		{
+			if (_queue.size() < _queue_max_size) {
+				_queue.push(evt);
+				return error_code::success;
+			}
+
+			//report errors
+			api_status::try_update(status, error_code::background_queue_overflow, "dropped event: " + evt);
+			return error_code::background_queue_overflow;
+		}
+
+	private:
+		void timer()//the timer triggers a queue flush (run in background)
+		{
+			while (_thread_is_running)
+			{
+				std::this_thread::sleep_for(std::chrono::milliseconds(_batch_timeout_ms));
+				flush();
+			}
+		}
+
+		void flush()//flush all batches
+		{
+		  const auto queue_size = _queue.size();
+			if (queue_size == 0) return;
+
+			//handle batching
+			std::string batch, next;
+			_queue.pop(&batch); //there is at least one element
+
+			for (size_t i = 1; i < queue_size; ++i)
+			{
+				_queue.pop(&next);
+
+				size_t batch_size = batch.length() + next.length();
+				if (batch_size > _batch_max_size)
+				{
+          api_status status;
+					//the batch is about to reach the max size: send it
+					if(_sender.send(batch,&status) != error_code::success)
+					{
+            if (_perror_cb) _perror_cb->report_error(status);
+					}
+          
+					batch = next;
+				}
+				else
+					batch += "\n" + next;
```

Comment:
>batch += "\n" + next; [](start = 5, length = 21)

switch to ostrstream to avoid copying and re-allocation?

Reply:
Yes, I'm worried about the time complexity here.  This: https://stackoverflow.com/questions/15082170/c-stdstring-append-vs-push-back seems to suggest that append amortizes to the additional bytes required which is reasonablish.  However, you might be creating a string and then adding it, which is doubling up on the work.  
Using an ostrstream does seem like it might work better.

Even better would be using some of the techniques in VW.   Instead of a queue, have a char* ring buffer and modify serialize to write directly into the ring buffer.  The modify flush() to just write an extent.  Locking would apply to the pointers indicating beginning and end of the queue. 

Even better would of course be sending bytes instead of strings...

Reply:
Switch to ostrstream and add client lib work item to switch to bytes?  Probably makes sense to schedule that with moving to a binary format?

---
In reply to: [187109810](https://github.com/JohnLangford/vowpal_wabbit/pull/1479#discussion_r187109810) [](ancestors = 187109810)

Reply:
That seems plausible.  Note that switching to ostrstream isn't quite enough---you want to have it defined at the class level not the function level so as to reuse memory.

Reply:
Makes sense.  Will need to consider concurrency.  

---
In reply to: [187339305](https://github.com/JohnLangford/vowpal_wabbit/pull/1479#discussion_r187339305) [](ancestors = 187339305)